src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -252,6 +455,270 @@ void symbol_factoryt::gen_nondet_array_init(
   }
 }
 
+symbolt &
+symbol_factoryt::new_tmp_symbol(const typet &type, const std::string &prefix)
+{
+  auto &symbol = get_fresh_aux_symbol(
+    type,
+    id2string(object_factory_params.function_id),
+    prefix,
+    loc,
+    ID_C,
+    symbol_table);
+  symbols_created.push_back(&symbol);
+  return symbol;
+}
+
+void symbol_factoryt::defer_size_initialization(
+  irep_idt associated_size_name,
+  irep_idt array_size_name)
+{
+  auto succeeded =
+    deferred_array_sizes.insert({associated_size_name, array_size_name});
+  INVARIANT(
+    succeeded.second,
+    "each size parameter should have a unique associated array");
+}
+
+optionalt<dstringt>
+symbol_factoryt::get_deferred_size(irep_idt symbol_name) const
+{
+  return optional_lookup(deferred_array_sizes, symbol_name);
+}
+
+void symbol_factoryt::gen_array_allocation(
+  code_blockt &assignments,
+  const exprt &array_expr,
+  const exprt &size)
+{
+  PRECONDITION(array_expr.type().id() == ID_pointer);
+  irep_idt malloc_symbol_name = CPROVER_PREFIX "malloc";
+  auto const *malloc_symbol = symbol_table.lookup(malloc_symbol_name);
+  if(!malloc_symbol)
+  {
+    malloc_symbol = &gen_malloc_function(malloc_symbol_name);
+  }
+  auto const &element_type = array_expr.type().subtype();
+  const exprt &array_size = size_of_expr(array_typet{element_type, size}, ns);
+
+  // array = malloc(sizeof(array[size]))
+  auto allocate_array =
+    side_effect_expr_function_callt{malloc_symbol->symbol_expr(),
+                                    exprt::operandst{array_size},
+                                    array_expr.type(),
+                                    loc};
+  assignments.add(code_assignt{array_expr, allocate_array});
+}
+
+void symbol_factoryt::gen_nondet_array_member_initialization(
+  code_blockt &assignments,
+  const exprt &array,
+  const exprt &array_size,
+  std::size_t depth,
+  const symbol_factoryt::recursion_sett &recursion_set)
+{
+  // for(size_t ix = 0; ix < size; ++ix) {
+  //   arr[ix] = nondet_init_T();
+  // }
+
+  auto const &array_index_symbol = new_tmp_symbol(size_type(), "index");
+  auto array_member_init = code_fort{};
+
+  array_member_init.init() = code_assignt{array_index_symbol.symbol_expr(),
+                                          from_integer(0, size_type())};
+
+  array_member_init.cond() = binary_exprt{
+    array_index_symbol.symbol_expr(), ID_lt, array_size, bool_typet{}};
+
+  auto array_member_init_body = code_blockt{};
+  gen_nondet_init(
+    array_member_init_body,
+    dereference_exprt{
+      plus_exprt{array, array_index_symbol.symbol_expr(), array.type()}},
+    depth,
```

Comment:
I think we need to pass in `depth + 1` here since we dereference the expression (since we pass in a `dereference_exprt`).